File: examples/hopping/mc.cpp
#include <entropy++/Csv.h>
#include <entropy++/sparse/MC_MI.h>
#include <entropy++/sparse/MC_W.h>
#include <entropy++/sparse/state/MC_W.h>
#include <entropy++/sparse/state/MC_MI.h>

#include <string>
#include <iostream>
#include <math.h>

#include <boost/filesystem.hpp>

using namespace std;
using namespace boost::filesystem;
using namespace entropy::sparse;
using namespace entropy::sparse::state;

# define MAX(a,b)    ((a>b)?a:b)
# define MAX3(a,b,c) MAX(a,MAX(b,c))

# define MIN(a,b)    ((a<b)?a:b)
# define MIN3(a,b,c) MIN(a,MIN(b,c))

# define POS 0
# define VEL 1
# define ACC 2
# define ACT 3
# define MSI 4 // muscle sensor input

// position      = SimData(:,2);
// velocity      = SimData(:,3);
// accelaration  = SimData(:,4);
// action        = SimData(:,10);
// muscle_sensor = SimData(:,5);

# define POS_MATLAB_INDEX                 (2  - 1)
# define VEL_MATLAB_INDEX                 (3  - 1)
# define ACC_MATLAB_INDEX                 (4  - 1)
# define ACT_MATLAB_INDEX                 (10 - 1)
# define MUSCLE_SENSOR_INPUT_MATLAB_INDEX (5  - 1)

int main(int argc, char **argv)
{
  if(argc != 2)
  {
    cout << "usage: " << argv[0] << " <directory>" << endl;
    exit(-1);
  }

  string d = string(argv[1]);

  Csv *csv = new Csv();

  Container *dcmot  = csv->read(d + "/dcmot.csv",  4,
                                POS_MATLAB_INDEX,
                                VEL_MATLAB_INDEX,
                                ACC_MATLAB_INDEX,
                                ACT_MATLAB_INDEX);

  Container *muslin = csv->read(d + "/muslin.csv", 5,
                                POS_MATLAB_INDEX,
                                VEL_MATLAB_INDEX,
                                ACC_MATLAB_INDEX,
                                ACT_MATLAB_INDEX,
                                MUSCLE_SENSOR_INPUT_MATLAB_INDEX);

  Container *musfib = csv->read(d + "/musfib.csv", 5,
                                POS_MATLAB_INDEX,
                                VEL_MATLAB_INDEX,
                                ACC_MATLAB_INDEX,
                                ACT_MATLAB_INDEX,
                                MUSCLE_SENSOR_INPUT_MATLAB_INDEX);

  Container *dcmotW  = dcmot->columns(3, 0, 1, 2);
  Container *dcmotS  = dcmot->columns(2, 0, 1);
  Container *dcmotA  = dcmot->columns(1, 3);
  delete dcmot;

  Container *muslinW = muslin->columns(3, 0, 1, 2);
  Container *muslinS = muslin->columns(1, 4);
  Container *muslinA = muslin->columns(1, 3);
  delete muslin;

  Container *musfibW = musfib->columns(3, 0, 1, 2);
  Container *musfibS = musfib->columns(1, 4);
  Container *musfibA = musfib->columns(1, 3);
  delete musfib;

  double p_min = MIN3(dcmotW->min(0), muslinW->min(0), musfibW->min(0));
  double p_max = MAX3(dcmotW->max(0), muslinW->max(0), musfibW->max(0));

  double v_min = MIN3(dcmotW->min(1), muslinW->min(1), musfibW->min(1));
  double v_max = MAX3(dcmotW->max(1), muslinW->max(1), musfibW->max(1));

  double a_min = MIN3(dcmotW->min(2), muslinW->min(2), musfibW->min(2));
  double a_max = MAX3(dcmotW->max(2), muslinW->max(2), musfibW->max(2));

  double mi_min = MIN(muslinS->min(0), musfibS->min(0));
  double mi_max = MAX(muslinS->max(0), musfibS->max(0));

  double ac_min = dcmotA->min(2);
  double ac_max = dcmotA->max(2);

  cout << "Domains:" << endl;
  cout << "  Position:          " << p_min  << " " << p_max  << endl;
  cout << "  Velocity:          " << v_min  << " " << v_max  << endl;
  cout << "  Acceleration:      " << a_min  << " " << a_max  << endl;
  cout << "  Action (DC Motor): " << ac_min << " " << ac_max << endl;

  // normalising DCMot
  dcmotW->normaliseColumn(0, p_min,  p_max);
  dcmotW->normaliseColumn(1, v_min,  v_max);
  dcmotW->normaliseColumn(2, a_min,  a_max);

  dcmotS->normaliseColumn(0, p_min,  p_max);
  dcmotS->normaliseColumn(1, v_min,  v_max);

  dcmotA->normaliseColumn(0, ac_min, ac_max);

  // normalising MusLin
  muslinW->normaliseColumn(0, p_min,  p_max);
  muslinW->normaliseColumn(1, v_min,  v_max);
  muslinW->normaliseColumn(2, a_min,  a_max);

  muslinS->normaliseColumn(0, mi_min, mi_max);

  // normalising MusFib
  musfibW->normaliseColumn(0, p_min,  p_max);
  musfibW->normaliseColumn(1, v_min,  v_max);
  musfibW->normaliseColumn(2, a_min,  a_max);

  musfibS->normaliseColumn(0, mi_min, mi_max);

  double **wdomain = new double*[3];
  for(int i = 0; i < 3; i++)
  {
    wdomain[i] = new double[2];
    wdomain[i][0] = 0.0;
    wdomain[i][1] = 1.0;
  }
  int bins   = 300;
  int *wbins = new int[3];
  wbins[0]   = bins;
  wbins[1]   = bins;
  wbins[2]   = bins;

  double **sdomain = new double*[2];
  for(int i = 0; i < 2; i++)
  {
    sdomain[i] = new double[2];
    sdomain[i][0] = 0.0;
    sdomain[i][1] = 1.0;
  }
  int *sbins = new int[2];
  sbins[0]   = bins;
  sbins[1]   = bins;

  double **adomain = new double*[1];
  for(int i = 0; i < 1; i++)
  {
    adomain[i] = new double[2];
    adomain[i][0] = 0.0;
    adomain[i][1] = 1.0;
  }
  int *abins = new int[1];
  abins[0]   = bins;

  dcmotW->setDomains(wdomain);
  dcmotW->setBinSizes(wbins);
  dcmotS->setDomains(sdomain);
  dcmotS->setBinSizes(sbins);
  dcmotA->setDomains(adomain);
  dcmotA->setBinSizes(abins);

  muslinW->setDomains(wdomain);
  muslinW->setBinSizes(wbins);
  muslinS->setDomains(sdomain);
  muslinS->setBinSizes(sbins);
  muslinA->setDomains(adomain);
  muslinA->setBinSizes(abins);

  musfibW->setDomains(wdomain);
  musfibW->setBinSizes(wbins);
  musfibS->setDomains(sdomain);
  musfibS->setBinSizes(sbins);
  musfibA->setDomains(adomain);
  musfibA->setBinSizes(abins);

  cout << "discretising dcmot W" << endl;
  Container *DdcmotW  = dcmotW->discretise(); delete dcmotW;
  cout << "discretising dcmot S" << endl;
  Container *DdcmotS  = dcmotS->discretise(); delete dcmotS;
  cout << "discretising dcmot A" << endl;
  Container *DdcmotA  = dcmotA->discretise(); delete dcmotA;
  Container *dcW2 = DdcmotW->drop(1);
  Container *dcW1 = DdcmotW->drop(-1);
  Container *dcA1 = DdcmotA->drop(-1);
  Container *dcS1 = DdcmotS->drop(-1);

  cout << "discretising muslin W" << endl;
  Container *DmuslinW  = muslinW->discretise(); delete muslinW;
  cout << "discretising muslin S" << endl;
  Container *DmuslinS  = muslinS->discretise(); delete muslinS;
  cout << "discretising muslin A" << endl;
  Container *DmuslinA  = muslinA->discretise(); delete muslinA;
  Container *mlW2 = DmuslinW->drop(1);
  Container *mlW1 = DmuslinW->drop(-1);
  Container *mlA1 = DmuslinA->drop(-1);
  Container *mlS1 = DmuslinS->drop(-1);

  cout << "discretising musfib W" << endl;
  Container *DmusfibW  = musfibW->discretise(); delete musfibW;
  cout << "discretising musfib S" << endl;
  Container *DmusfibS  = musfibS->discretise(); delete musfibS;
  cout << "discretising musfib A" << endl;
  Container *DmusfibA  = musfibA->discretise(); delete musfibA;
  Container *mfW2 = DmusfibW->drop(1);
  Container *mfW1 = DmusfibW->drop(-1);
  Container *mfA1 = DmusfibA->drop(-1);
  Container *mfS1 = DmusfibS->drop(-1);

  cout << "Calculating DCMot:" << endl;
  double     dcmc_w   = entropy::sparse::MC_W(dcW2, dcW1, dcA1);
  double     dcmc_mi  = entropy::sparse::MC_MI(dcW2, dcW1, dcS1, dcA1);
  Container* dcmc_mid = entropy::sparse::state::MC_MI(dcW2, dcW1, dcS1, dcA1);
  Container* dcmc_wd  = entropy::sparse::state::MC_W(dcW2, dcW1, dcA1);
  double     s        = dcmc_mid->columnSum(0) / ((double)dcmc_mid->rows());
  double     ss       = dcmc_wd->columnSum(0)  / ((double)dcmc_mid->rows());
  cout << "  DCMot MC_W:  " << dcmc_w  << endl;
  cout << "  DCMot MC_MI: " << dcmc_mi << endl;
  cout << "  checking values: " << s  << " vs. " << dcmc_mi << " = " << fabs(s  - dcmc_mi) << endl;
  cout << "  checking values: " << ss << " vs. " << dcmc_w  << " = " << fabs(ss - dcmc_w)  << endl;
  cout << "writing DCMot state-dependent MC_W" << endl;
  csv->write("state-dependent-dc_mc_w.csv", dcmc_wd);
  cout << "writing DCMot state-dependent MC_MI" << endl;
  csv->write("state-dependent-dc_mc_mi.csv", dcmc_mid);

  cout << "Calculating MusLin:" << endl;
  double     mlmc_w   = entropy::sparse::MC_W(mlW2, mlW1, mlA1);
  double     mlmc_mi  = entropy::sparse::MC_MI(mlW2, mlW1, mlS1, mlA1);
  Container* mlmc_mid = entropy::sparse::state::MC_MI(mlW2, mlW1, mlS1, mlA1);
  Container* mlmc_wd  = entropy::sparse::state::MC_W(mlW2, mlW1, mlA1);
  double     t        = mlmc_mid->columnSum(0) / ((double)mlmc_mid->rows());
  double     tt       = mlmc_wd->columnSum(0) / ((double)mlmc_wd->rows());
  cout << "MusLin MC_W:  " << mlmc_w  << endl;
  cout << "MusLin MC_MI: " << mlmc_mi << endl;
  cout << "  checking values: " << t  << " vs. " << mlmc_mi << " = " << fabs(t  - mlmc_mi) << endl;
  cout << "  checking values: " << tt << " vs. " << mlmc_w  << " = " << fabs(tt - mlmc_w)  << endl;
  cout << "writing MusLin state-dependent MC_W" << endl;
  csv->write("state-dependent-ml_mc_w.csv", mlmc_wd);
  cout << "writing MusLin state-dependent MC_MI" << endl;
  csv->write("state-dependent-ml_mc_mi.csv", mlmc_mid);

  cout << "Calculating MusFib:" << endl;
  double     mfmc_w   = entropy::sparse::MC_W(mfW2, mfW1, mfA1);
  double     mfmc_mi  = entropy::sparse::MC_MI(mfW2, mfW1, mfS1, mfA1);
  Container* mfmc_mid = entropy::sparse::state::MC_MI(mfW2, mfW1, mfS1, mfA1);
  Container* mfmc_wd  = entropy::sparse::state::MC_W(mfW2, mfW1, mfA1);
  double     u        = mfmc_mid->columnSum(0) / ((double)mfmc_mid->rows());
  double     uu       = mfmc_wd->columnSum(0)  / ((double)mfmc_mid->rows());
  cout << "MusFib MC_W:  " << mfmc_w  << endl;
  cout << "MusFib MC_MI: " << mfmc_mi << endl;
  cout << "  checking values: " << u  << " vs. " << mfmc_mi << " = " << fabs(u  - mfmc_mi) << endl;
  cout << "  checking values: " << uu << " vs. " << mfmc_w  << " = " << fabs(uu - mfmc_w)  << endl;
  cout << "writing MusFib state-dependent MC_W" << endl;
  csv->write("state-dependent-mf_mc_w.csv", mfmc_wd);
  cout << "writing MusFib state-dependent MC_MI" << endl;
  csv->write("state-dependent-mf_mc_mi.csv", mfmc_mid);

}

File: examples/hopping/mc.py
#!/usr/bin/env python

import sys
sys.path.append("../../py/")

from entropy import container, mc_w 
from entropy.state import mc_w as mc_w_d

fd = open("data/dcmot.csv","r")
data = fd.readlines()
fd.close()
dcmot = []
for line in data:
    values = [float(v) for v in line.split(",")]
    dcmot.append([values[1], values[2], values[3], values[9]])

fd = open("data/muslin.csv","r")
data = fd.readlines()
fd.close()
muslin = []
for line in data:
    values = [float(v) for v in line.split(",")]
    muslin.append([values[1], values[2], values[3], values[9], values[4]])

fd = open("data/musfib.csv","r")
data = fd.readlines()
fd.close()
musfib = []
for line in data:
    values = [float(v) for v in line.split(",")]
    musfib.append([values[1], values[2], values[3], values[9], values[4]])

dcmotW = container.Container(len(dcmot), 3)
dcmotS = container.Container(len(dcmot), 2)
dcmotA = container.Container(len(dcmot), 1)

muslinW = container.Container(len(muslin), 3)
muslinS = container.Container(len(muslin), 1)
muslinA = container.Container(len(muslin), 1)

musfibW = container.Container(len(musfib), 3)
musfibS = container.Container(len(musfib), 1)
musfibA = container.Container(len(musfib), 1)

for i in range(0, len(dcmot)):
    dcmotW.add(dcmot[i][0])
    dcmotW.add(dcmot[i][1])
    dcmotW.add(dcmot[i][2])
    dcmotS.add(dcmot[i][0])
    dcmotS.add(dcmot[i][1])
    dcmotA.add(dcmot[i][3])

for i in range(0, len(muslin)):
    muslinW.add(muslin[i][0])
    muslinW.add(muslin[i][1])
    muslinW.add(muslin[i][2])
    muslinS.add(muslin[i][4])
    muslinA.add(muslin[i][3])

for i in range(0, len(musfib)):
    musfibW.add(musfib[i][0])
    musfibW.add(musfib[i][1])
    musfibW.add(musfib[i][2])
    musfibS.add(musfib[i][4])
    musfibA.add(musfib[i][3])

p_min = min(dcmotW.data[:,0].min(), muslinW.data[:,0].min(),
        musfibW.data[:,0].min())
p_max = max(dcmotW.data[:,0].max(), muslinW.data[:,0].max(),
        musfibW.data[:,0].max())

v_min = min(dcmotW.data[:,1].min(), muslinW.data[:,1].min(),
        musfibW.data[:,1].min())
v_max = max(dcmotW.data[:,1].max(), muslinW.data[:,1].max(),
        musfibW.data[:,1].max())

a_min = min(dcmotW.data[:,2].min(), muslinW.data[:,2].min(),
        musfibW.data[:,2].min())
a_max = max(dcmotW.data[:,2].max(), muslinW.data[:,2].max(),
        musfibW.data[:,2].max())

mi_min = min(muslinS.data[:,0].min(), musfibS.data[:,0].min())
mi_max = max(muslinS.data[:,0].max(), musfibS.data[:,0].max())

ac_min = dcmotA.data[:,0].min()
ac_max = dcmotA.data[:,0].max()

print "Domains:"
print "  Position:          " + str(p_min)  + " " + str(p_max)
print "  Velocity:          " + str(v_min)  + " " + str(v_max)
print "  Acceleration:      " + str(a_min)  + " " + str(a_max)
print "  Action (DC Motor): " + str(ac_min) + " " + str(ac_max)


dcmotW.normaliseColumn(0,  p_min,  p_max)
dcmotW.normaliseColumn(1,  v_min,  v_max)
dcmotW.normaliseColumn(2,  a_min,  a_max)
dcmotS.normaliseColumn(0,  p_min,  p_max)
dcmotS.normaliseColumn(1,  v_min,  v_max)
dcmotA.normaliseColumn(0,  ac_min, ac_max)

dcmotW.setBins([300, 300, 300])
dcmotW.setDomains([ (0.0, 1.0), (0.0, 1.0), (0.0, 1.0)])
dcmotS.setBins([300, 300])
dcmotS.setDomains([ (0.0, 1.0), (0.0, 1.0)])
dcmotA.setBins([300])
dcmotA.setDomains([ (0.0, 1.0) ])


muslinW.normaliseColumn(0, p_min,  p_max)
muslinW.normaliseColumn(1, v_min,  v_max)
muslinW.normaliseColumn(2, a_min,  a_max)
muslinS.normaliseColumn(0, mi_min, mi_max)

muslinW.setBins([300, 300, 300])
muslinW.setDomains([ (0.0, 1.0), (0.0, 1.0), (0.0, 1.0)])
muslinS.setBins([300, 300])
muslinS.setDomains([ (0.0, 1.0), (0.0, 1.0)])
muslinA.setBins([300])
muslinA.setDomains([ (0.0, 1.0) ])


musfibW.setBins([300, 300, 300])
musfibW.setDomains([ (0.0, 1.0), (0.0, 1.0), (0.0, 1.0)])
musfibS.setBins([300, 300])
musfibS.setDomains([ (0.0, 1.0), (0.0, 1.0)])
musfibA.setBins([300])
musfibA.setDomains([ (0.0, 1.0) ])
musfibW.normaliseColumn(0, p_min,  p_max)
musfibW.normaliseColumn(1, v_min,  v_max)
musfibW.normaliseColumn(2, a_min,  a_max)
musfibS.normaliseColumn(0, mi_min, mi_max)


dcmotW.discretise()
dcmotS.discretise()
dcmotA.discretise()

muslinW.discretise()
muslinS.discretise()
muslinA.discretise()

musfibW.discretise()
musfibS.discretise()
musfibA.discretise()

dcmotW2 = dcmotW.drop(1)
dcmotW1 = dcmotW.drop(-1)
dcmotA1 = dcmotA.drop(-1)

muslinW2 = muslinW.drop(1)
muslinW1 = muslinW.drop(-1)
muslinA1 = muslinA.drop(-1)

musfibW2 = musfibW.drop(1)
musfibW1 = musfibW.drop(-1)
musfibA1 = musfibA.drop(-1)

mc_w_dcmot  = mc_w.MC_W(dcmotW2,  dcmotW1,  dcmotA1)
mc_w_muslin = mc_w.MC_W(muslinW2, muslinW1, muslinA1)
mc_w_musfib = mc_w.MC_W(musfibW2, musfibW1, musfibA1)

print "MC_W: "
print " dcmot:  " + str(mc_w_dcmot)
print " muslin: " + str(mc_w_muslin)
print " musfib: " + str(mc_w_musfib)

mc_w_dcmot_state  = mc_w_d.MC_W(dcmotW2,  dcmotW1,  dcmotA1)
mc_w_muslin_state = mc_w_d.MC_W(muslinW2, muslinW1, muslinA1)
mc_w_musfib_state = mc_w_d.MC_W(musfibW2, musfibW1, musfibA1)

print "MC_W: "
print " dcmot:  " + str(mc_w_dcmot_state)
print " muslin: " + str(mc_w_muslin_state)
print " musfib: " + str(mc_w_musfib_state)

fd = open("dcmot.csv","w")
for i in range(0, mc_w_dcmot_state.shape[0]):
    fd.write(str(mc_w_dcmot_state[i]) + "\n")
fd.close()

fd = open("muslin.csv","w")
for i in range(0, mc_w_muslin_state.shape[0]):
    fd.write(str(mc_w_muslin_state[i]) + "\n")
fd.close()

fd = open("musfib.csv","w")
for i in range(0, mc_w_musfib_state.shape[0]):
    fd.write(str(mc_w_musfib_state[i]) + "\n")
fd.close()

